### PR TITLE
[WIP] tpm2_import: new tool that uses tpm2_import to import a key

### PR DESCRIPTION
**Pull request by williamcroberts**
TODO:
1. --parent-key-public should be optional, just read it in if not specified.
2. I don't understand the difference between the imported private, public and
   symetric. Is this limited to just importing an AES-128 key? If so why do I
   need a private and public key?
3. Should the mandatory inputs be options or arguments?

Limited to importing a symmetric AES-128 key.
^^^
Is this true?

Add tests.

Signed-off-by: William Roberts <william.c.roberts@intel.com>

### CODE REVIEW

**Comment by williamcroberts** [comment]
@idesai can you help answer this?

I don't understand the difference between the imported private, public and
symetric. Is this limited to just importing an AES-128 key? If so why do I
need a private and public key?

What would we need to do to import private/pub key pairs if it doesn't support it now? Is this possible?

**Comment by idesai** [comment]
@williamcroberts import_priv and import_pub are describing the sensitive and public information about the symmetric key. Another place you can see this is also tpm2_create tool when used to create a symmetric key produces private(encrypted sensitive) and public(key description). You can then use this in tpm2_load tool that needs both the encrypted sensitive and public data to produce a useable key handle.

**Comment by idesai** [comment]
@williamcroberts a test is added part of the commit  test/system/test_tpm2_import.sh

**Comment by williamcroberts** [comment]
@idesai Ahh when discussing private and public were talking about the tpm's notion of a public and private portions of a key object. This is easy to confuse with private and public portions of a key.

Ok I see now that the symmetric AES 128 key  is the one being imported, and the public/private are output that can be used in load later. I can make the manpage more clear about this.

So we would probably want the ability to load all supported symmetric keys? Can tpm2_import asymmetric keys?

File: test/system/test_tpm2_import.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+# THE POSSIBILITY OF SUCH DAMAGE.
+#;**********************************************************************;
+#!/bin/sh
+
```
Comment (williamcroberts) [inline test/system/test_tpm2_import.sh]: Update to use traps for handling file cleanup (to ensure that it is done). Also use ERR trap to print the line number so we don't have to explicitly check return codes.

File: test/system/test_tpm2_import.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+rm -f parent.ctx parent.pub  plain.dec.ssl  plain.enc  plain.txt  sym.key
+
+tpm2_createprimary -G 1 -g 0xb -A o -C parent.ctx 
+tpm2_evictcontrol -A o -c parent.ctx -S 0x81010005 
+
+dd if=/dev/urandom of=sym.key bs=1 count=16
+
+tpm2_readpublic -H 0x81010005 --opu parent.pub
+
+tpm2_import -k sym.key -H 0x81010005 -f parent.pub -q import_key.pub \
```
Comment (williamcroberts) [inline test/system/test_tpm2_import.sh]: make -f optional

Comment (williamcroberts) [reply test/system/test_tpm2_import.sh]: How would -q work with an asymmetric key?

File: test/system/test_tpm2_import.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+-r import_key.priv
+
+if [ $? != 0 ];then
+	    echo "tpm2_import Failed"
+		exit 1
+fi
+tpm2_load  -H 0x81010005 -u import_key.pub -r import_key.priv -n import_key.name \
+-C import_key.ctx
```
Comment (williamcroberts) [inline test/system/test_tpm2_import.sh]: I'm just noticing outputs unused, likely tpm2_load if not given a -n option can just not output the name, or just send it to stdout...

File: test/system/test_tpm2_import.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+openssl enc -in plain.enc -out plain.dec.ssl -d -K `xxd -p sym.key` -iv 0 \
+-aes-128-cfb
+
+diff plain.txt plain.dec.ssl
+if [ $? != 0 ];then
+	echo "tpm2_import test failed"
+	exit 1
+fi
+
+tpm2_evictcontrol -A o -H 0x81010005 -S 0x81010005
```
Comment (williamcroberts) [inline test/system/test_tpm2_import.sh]: I also see maybe a case to make tpm2_evictcontrol smart enough that if assume -H if -S is not provided, or maybe vice versa?

**Comment by williamcroberts** [comment]
I don't understand why version 1.0.2h of openssl is missing this routine:
```
../tools/tpm2_import.c: In function ‘encrypt_seed_with_tpm2_rsa_public_key’:
../tools/tpm2_import.c:135:5: error: implicit declaration of function ‘RSA_padding_add_PKCS1_OAEP_mgf1’ [-Werror=implicit-function-declaration]
     int return_code = RSA_padding_add_PKCS1_OAEP_mgf1(encoded,
```

**Comment by williamcroberts** [comment]
Darn, it's really version 1.0.1f according to the apt-get install output: 1.0.1f-1ubuntu2.22

Grabing the source for the package from: https://launchpad.net/ubuntu/+source/openssl/1.0.1f-1ubuntu2.22

untaring it and searching yields nothing:
```
grep -rn RSA_padding_add_PKCS1_OAEP_mgf1
```

**Comment by idesai** [comment]
Dang! openssl hmm.... @williamcroberts when I wrote this app, the ssl version on my system was: OpenSSL 1.0.2g  1 Mar 2016.

File: test/system/test_tpm2_import.sh (hunk 3 above)
Comment (idesai) [reply test/system/test_tpm2_import.sh]: @williamcroberts, agreed. If not given the -n option it should just print it to stdout.

File: test/system/test_tpm2_import.sh (hunk 2 above)
Comment (idesai) [reply test/system/test_tpm2_import.sh]: @williamcroberts I believe the unique data should have the asymmetric public key.

**Comment by idesai** [comment]
@williamcroberts yup that's the TPM's notion of sensitive and public data, yeah that can be confusing :)

File: test/system/test_tpm2_import.sh (hunk 4 above)
Comment (idesai) [reply test/system/test_tpm2_import.sh]: @williamcroberts agreed. it is not very intuitive when trying to remove the persistent handle.

**Comment by williamcroberts** [comment]
@idesai my system had the same openssl version. The configure script handles the version check, so I see no need to mention that in the manpage.

**Comment by idesai** [comment]
@williamcroberts will this be merged?

**Comment by williamcroberts** [comment]
@idesai yes, Ill try and do it soon.

**Comment by williamcroberts** [comment]
oh thank god for restore branch... got a little overzealous with my house keeping.

**Comment by williamcroberts** [comment]
@martinezjavier is the version bump of openssl going to affect you when you go to package? This change requires: 1.0.2g

**Comment by martinezjavier** [comment]
@williamcroberts at least for Fedora it won't be an issue, since the last version of the compat-openssl10 package is 1.0.2j. I don't know for the other distros.

Did you check if these API change in OpenSSL 1.1.x? I fixed the issues we had on commit 5cee30cbc3da ("kdfa: allow to build with OpenSSL 1.1.x"), it would be nice if we make sure that no regressions are introduced (I didn't check BTW, just mentioning in case).

**Comment by idesai** [comment]
@williamcroberts, cool this branch is back online, i was looking for it :) 
did I miss any questions? or is it ready to be merged?

**Comment by idesai** [comment]
@williamcroberts made the changes per TSS commit: libmu: convert names to comply with TCG spec.

**Comment by martinezjavier** [comment]
> Did you check if these API changed in OpenSSL 1.1.x? I fixed the issues we had on commit 5cee30c ("kdfa: allow to build with OpenSSL 1.1.x"), it would be nice if we make sure that no regressions are introduced (I didn't check BTW, just mentioning in case).

@williamcroberts so this branch added a build error with OpenSSL 1.1.0...

```
tools/tpm2_import.c:99:11: error: dereferencing pointer to incomplete type ‘RSA {aka struct rsa_st}’
     if ((r->n == NULL && n == NULL) || (r->e == NULL && e == NULL)) {
           ^~
make: *** [Makefile:1858: tools/tpm2_import.o] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by martinezjavier** [comment]
I created PR #563 that fixes this issue among others I found while reading the code.

File: .travis.yml [hunk 5]
```diff
@@ -65,6 +63,13 @@ install:
     - sha256sum libcmocka0_1.0.1-2_amd64.deb | grep -q 797155b45a8288a860c4ed9dd3f161420f09ebf362de30166d9f6b98bfc27dd0
     - sudo dpkg -i libcmocka0_1.0.1-2_amd64.deb
     - sudo dpkg -i libcmocka-dev_1.0.1-2_amd64.deb
+    # openssl 1.0.2g
+    - wget http://security.ubuntu.com/ubuntu/pool/main/o/openssl/libssl1.0.0_1.0.2g-1ubuntu4.6_amd64.deb
+    - wget http://security.ubuntu.com/ubuntu/pool/main/o/openssl/libssl-dev_1.0.2g-1ubuntu4.6_amd64.deb
+    - sha256sum libssl1.0.0_1.0.2g-1ubuntu4.6_amd64.deb | grep -q b858321c5858533399320fc2d849962c25288fe17f545dc4fb3ce99630cea5fe
+    - sha256sum libssl-dev_1.0.2g-1ubuntu4.6_amd64.deb | grep -q e4ad54c12aaf88d65c04eaccf1e07baf81ab3d14acf7e118137e8b1e10bdd06b
+    - sudo dpkg -i libssl1.0.0_1.0.2g-1ubuntu4.6_amd64.deb
+    - sudo dpkg -i libssl-dev_1.0.2g-1ubuntu4.6_amd64.deb
```
Comment (liuqun) [inline .travis.yml]: It seemd to be a typo mistake that the commit message e63b52b said version 1.0.1g, which should be 1.0.2g

**Pull request merged**